### ISSUE TITLE
Overflowed stack

### ISSUE THREAD

**Issue by JustForFun88** [rust-lang/rust#105656]
I tried this [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=53041ede576d53f601caa914317a2ff1):

```rust
struct CharBuffer {
    buf: [u8; 4],
    len: usize,
}

impl From<CharBuffer> for char {
    #[inline]
    fn from(char: CharBuffer) -> Self {
        char.into()
    }
}

fn main() {
    let foo = CharBuffer{ buf: [0; 4], len: 0 };
    let bar: char = char::from(foo);
}
```

I expected to see this happen:
```
error[E0277]: the trait bound `char: From<CharBuffer>` is not satisfied
 --> src/main.rs:9:9
  |
9 |         char.into()
  |         ^^^^ ---- required by a bound introduced by this call
  |         |
  |         the trait `From<CharBuffer>` is not implemented for `char`
  |
  = note: required for `CharBuffer` to implement `Into<char>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

Instead, this happened:
```
thread 'main' has overflowed its stack
```

**Comment by bjorn3** [comment]
I can reproduce this on the playground when using nightly, but not locally.

**Comment by albertlarsan68** [comment]
It is the created binary that crashes, not the rustc process.

**Comment by bjorn3** [comment]
That explains why I couldn't reproduce it locally :sweat_smile: I didn't run the executable. Only rustc.

**Comment by JustForFun88** [comment]
I know that in general it is a problem of recursive call. I was just confused by the silent compilation.

The compiler didn't even warn that there was a recursive call of the same function.

**Comment by albertlarsan68** [comment]
The bug may be this, that there is no diagnostic concerning this unconditional recursion.

@rustbot label +A-diagnostics +D-terse

**Comment by quat1024** [comment]
This is a case of https://github.com/rust-lang/rust/issues/57965, the `unconditional_recursion` lint not detecting mutual recursion.

**Comment by jruderman** [comment]
I think this can be closed as a duplicate of #98217
